### PR TITLE
feat: add gRPC client for git-sensor

### LINKED ISSUE #3086: Feature: Add gRPC client for git sensor
### 🔖 Feature description

Currently, the communication with git sensor is done over REST APIs. To improve performance gRPC layer should be introduced.

### 🎤 Pitch

Performance improvement

### ✌️ Solution

Add gRPC server along side REST server

### 🔄️ Alternative

_No response_

### 👀 Have you spent some time to check if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/devtron-labs/devtron/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by vishal-dt**
# Description

Add gRPC client for git-sensor

Fixes #3086 

### New Environment variables introduced:
 - GIT_SENSOR_PROTOCOL -> possible values are REST and GRPC
 
Note:
-  For GRPC, do not include `http` in GIT_SENSOR_URL

## How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Demo cluster

## Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR requires documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code.
* [x] I have commented my code, particularly in hard-to-understand areas.
* [ ] I have tested it for all user roles.
* [x] I have added all the required unit/api test cases.

## Does this PR introduce a user-facing change?

NO
```release-note

```

### CODE REVIEW

File: client/gitSensor/GitSensorGrpcClient.go [hunk 1]
```diff
@@ -0,0 +1,710 @@
...
+		Name:          config.Name,
+		EventTypesCsv: config.EventTypesCsv,
+		ActionType:    config.ActionType,
+		IsActive:      config.IsActive,
+		Selectors:     selectors,
+	}
+	if !config.CreatedOn.IsZero() {
+		mappedConfig.CreatedOn = timestamppb.New(config.CreatedOn)
+	}
+	if !config.UpdatedOn.IsZero() {
```
Comment (kartik-579) [inline client/gitSensor/GitSensorGrpcClient.go:683]: we can extract this logic and create a new wrapper method

File: client/gitSensor/GitSensorRestClient.go [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+				return nil, err
+			}
+		}
+	}
+	client := &http.Client{Timeout: time.Duration(config.Timeout) * time.Second}
+	return &RestClientImpl{httpClient: client, logger: logger, baseUrl: baseUrl}, nil
+}
+
+func (session RestClientImpl) GetHeadForPipelineMaterials(ctx context.Context, req *HeadRequest) (material []*CiPipelineMaterial, err error) {
+	request := &ClientRequest{ResponseBody: &material, Method: "POST", RequestBody: req, Path: "git-head"}
```
Comment (kartik-579) [inline client/gitSensor/GitSensorRestClient.go:304]: we can extract method types as const

File: pkg/pipeline/CiHandler.go [hunk 3]
```diff
@@ -1004,16 +1007,16 @@ func (impl *CiHandlerImpl) getLastSeenCommit(ciMaterialId int) (bean.GitCommit,
...
 	if err != nil {
 		return bean.GitCommit{}, err
 	}
 	gitCommit := bean.GitCommit{
-		Commit:  hashResponse[0].GitCommit.Commit,
-		Author:  hashResponse[0].GitCommit.Author,
-		Date:    hashResponse[0].GitCommit.Date,
-		Message: hashResponse[0].GitCommit.Message,
-		Changes: hashResponse[0].GitCommit.Changes,
+		Commit:  res[0].GitCommit.Commit,
```
Comment (kartik-579) [inline pkg/pipeline/CiHandler.go:1015]: we can add a safety check for length of response

**Review by kartik-579 — CHANGES_REQUESTED** [review]

File: client/gitSensor/GitSensorGrpcClient.go [hunk 4]
```diff
@@ -0,0 +1,710 @@
...
+	defer cancel()
+
+	// Configure gRPC dial options
+	var opts []grpc.DialOption
+	opts = append(opts,
+		grpc.WithChainUnaryInterceptor(grpc_prometheus.UnaryClientInterceptor, otelgrpc.UnaryClientInterceptor()),
+		grpc.WithBlock(),
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithDefaultCallOptions(
+			grpc.MaxCallRecvMsgSize(MaxMsgSizeBytes),
```
[the remark this answers is not in the document]
Comment (vishal-dt) [reply client/gitSensor/GitSensorGrpcClient.go]: addressed

File: client/gitSensor/GitSensorGrpcClient.go [hunk 5]
```diff
@@ -0,0 +1,710 @@
...
+// AddRepo adds git materials
+func (client *GrpcApiClientImpl) AddRepo(ctx context.Context, materials []*GitMaterial) error {
+
+	serviceClient, err := client.getGitSensorServiceClient()
+	if err != nil {
+		return err
+	}
+
+	// Mapping req to proto type
+	gitMaterials := make([]*pb.GitMaterial, 0)
```
[the remark this answers is not in the document]
Comment (vishal-dt) [reply client/gitSensor/GitSensorGrpcClient.go]: addressed

File: client/gitSensor/GitSensorGrpcClient.go [hunk 6]
```diff
@@ -0,0 +1,710 @@
...
+	})
+	if err != nil {
+		return nil, err
+	}
+	if res == nil {
+		return nil, nil
+	}
+
+	// Mapping res
+	commits := make([]*GitCommit, 0)
```
[the remark this answers is not in the document]
Comment (vishal-dt) [reply client/gitSensor/GitSensorGrpcClient.go]: addressed

File: client/gitSensor/GitSensorGrpcClient.go (hunk 1 above)
Comment (vishal-dt) [reply client/gitSensor/GitSensorGrpcClient.go:680]: can be discussed

**Review by kartik-579 — APPROVED** [review]

**Pull request merged**